libsweep/src/arch/win/serial_win.cc

```diff
@@ -0,0 +1,280 @@
+#include "serial.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <windows.h>
+#include <tchar.h>
+
+namespace sweep {
+namespace serial {
+
+typedef struct error {
+  const char* what; // always literal, do not free
+} error;
+
+typedef struct device {
+  HANDLE fd;
+  OVERLAPPED ro_;
+  OVERLAPPED wo_;
+} device;
+
+// Constructor hidden from users
+static error_s error_construct(const char* what) {
+  SWEEP_ASSERT(what);
+
+  auto out = new error{what};
+  return out;
+}
+
+const char* error_message(error_s error) {
+  SWEEP_ASSERT(error);
+
+  return error->what;
+}
+
+void error_destruct(error_s error) {
+  SWEEP_ASSERT(error);
+
+  delete error;
+}
+
+static int32_t detail_get_port_number(const char* port, error_s* error) {
+  SWEEP_ASSERT(port);
+  SWEEP_ASSERT(error);
+
+  for (int i = 0; i < 99; i++) {
+    char str[10];
+    sprintf(str, "COM%d", i);
+    if (strcmp(port, str) == 0) {
+      return i;
+    }
+  }
```

Comment:
Hm this is a bit ugly. What about
- Checking for length > 3
- Calling `strtol` with [port + 3, last]
- Handling parsing errors

http://en.cppreference.com/w/cpp/string/byte/strtol